### PR TITLE
[lexicon] Define irrefutable and refutable pattern in the Lexicon.

### PR DESCRIPTION
**Pull request by gottesmm**
This term is used often times in SILGen/AST. This just defines the concept at a
high level in a centralized place.

### CODE REVIEW

**Comment by gottesmm** [comment]
@jckarter What do you think? I tried to not go into too much detail. Do you think I should add some more here?

**Comment by gottesmm** [comment]
@swift-ci smoke test

**Comment by gottesmm** [comment]
@swift-ci smoke test

**Comment by gottesmm** [comment]
Actually, if Joe wants me to change the language, I will change it post commit.

File: docs/Lexicon.rst [hunk 1]
```diff
@@ -144,6 +144,14 @@ source code, tests, and commit messages. See also the `LLVM lexicon`_.
     themselves. They can be compared across declarations but cannot be used
     directly from within the context.
 
+  irrefutable pattern
+    A pattern that always matches. These patterns either bind to a variable or
+    perform structural modification, e.x.:
+
+    1. ``case _:``.
+    2. ``case let x:``.
+    3. ``case (_, _):``.
```
Comment (jrose-apple) [inline docs/Lexicon.rst]: The last example can fail if the input is `Any`, but this wasn't my understanding of refutable vs. irrefutable patterns anyway. I thought it was entirely contextual—`let` patterns vs. `case` patterns.

Comment (jckarter) [inline docs/Lexicon.rst]: Currently, the compiler only statically accepts tuple patterns when the pattern type is a tuple. Perhaps we could relax that and implicitly insert an `as (_, _)` check, but they are irrefutable in the current model.

**Pull request merged**